keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2314 @@
+phase.
+
+![design](./design.svg)
+
+QoS-class resources are designed to be opaque to the CRI client in the sense
+that the container runtime takes care of configuration and control of the
+resources and the classes within.
+
+### Phase 1
+
+The goal is to enable a functional base for users to leverage QoS-class
+resources and start experimenting with them in Kubernetes:
+
+- extend the CRI protocol to allow QoS-class resource assignment and updates to
+  be communicated from kubelet to the runtime
+- extend the CRI protocol to allow runtime to communicate available QoS-class
+  resources (the types of resources and the classes within) to kubelet
+- introduce a feature gate for enabling QoS-class resource support in kubelet
+- extend PodSpec to support assignment of QoS-clsss resources
+- extend NodeStatus to show availability/capacity of QoS-clsss resources on a
+  node
+
+### Future work
+
+This section sheds light on the end goal of this work in order to better
+evaluate this KEP in a broader context. What a fully working solution would
+consists of and what the (next) steps to accomplish that would be. These topics
+are currently listed as "future work" in [Goals](#goals).
+
+In practice, the future work mostly consists of changes to the Kubernetes API
+and control plane components.
+
+All the sub-sections below should be considered as `<<[UNRESOLVED]>>`.
+
+#### In-place pod vertical scaling
+
+Properly integrating QoS-class resources to
+[In-place pod vertical scaling](#1287) will allow update of running containers.
+However, some QoS-class resources may be immutable, meaning the in-place
+updates are not possible.
+
+#### Access control
+
+This future step adds support for controlling the access to available QoS-class
+resources.
+
+Access control could be achieved e.g. by extending ResourceQuotaSpec which
+would implement restrictions based on the namespace.
+
+```diff
+ // ResourceQuotaSpec defines the desired hard limits to enforce for Quota.
+ type ResourceQuotaSpec struct {
+     // hard is the set of desired hard limits for each named resource.
+     Hard ResourceList
+     // A collection of filters that must match each object tracked by a quota.
+     // If not specified, the quota matches all objects.
+     Scopes []ResourceQuotaScope 
+     // scopeSelector is also a collection of filters like scopes that must match each
+     // object tracked by a quota but expressed using ScopeSelectorOperator in combination
+     // with possible values.
+     ScopeSelector *ScopeSelector
++   // QOSResources contains the desired set of allowed QoS resources.
++   // +optional
++   QOSResources QOSResourceQuota
+ }
+
++// QOSResourceQuota contains the allowed QoS resources.
++type QOSResourceQuota struct {
++       // Pod contains the allowed QoS resources for pods.
++       // +optional
++       Pod []AllowedQOSResource
++       // Container contains the allowed QoS resources for pods.
```

Comment:
Do you mean `Container contains the allowed QoS resources for containers.`?

Reply:
Yes, for containers. Thx for pointing this out. Fixed